homeassistant/components/backup/manager.py

```diff
@@ -0,0 +1,178 @@
+        """Generate a backup."""
+        if self.backing_up:
+            raise HomeAssistantError("Backup already in progress")
+
+        if self._backups is None:
+            self._backups = {}
+
+        try:
+            self.backing_up = True
+            backup_name = f"Core {VERSION}"
+            date_str = now().isoformat()
+            slug = _generate_slug(date_str, backup_name)
+
+            backup_data = {
+                "slug": slug,
+                "name": backup_name,
+                "date": date_str,
+                "type": "partial",
+                "folders": ["homeassistant"],
+                "homeassistant": {} if VERSION.dev else {"version": VERSION},
+                "compressed": True,
+            }
+            tar_file = Path(self.backup_dir, f"{slug}.tar")
+
+            if not self.backup_dir.exists():
+                LOGGER.debug("Creating backup directory")
+                self.hass.async_add_executor_job(self.backup_dir.mkdir)
+
+            def _create_backup() -> None:
+                with TemporaryDirectory() as tmp_dir:
+                    json_util.save_json(
+                        Path(tmp_dir, "backup.json").as_posix(),
+                        backup_data,
+                    )
+
+                    with tarfile.open(
+                        Path(tmp_dir, "homeassistant.tar.gz"), "w:gz"
+                    ) as tar:
+                        _add_directory_to_tarfile(
+                            tar_file=tar,
+                            origin_path=Path(self.hass.config.path()),
+                        )
+
+                    with tarfile.open(tar_file, "w:") as tar:
+                        tar.add(tmp_dir, arcname=".")
+
+            await self.hass.async_add_executor_job(_create_backup)
+            backup = Backup(
+                slug=slug,
+                name=backup_name,
+                date=date_str,
+                path=tar_file,
+                size=round(tar_file.stat().st_size / 1_048_576, 2),
+            )
+            self._backups[slug] = backup
+            LOGGER.debug("Generated new backup with slug %s", slug)
+            return backup
+        finally:
+            self.backing_up = False
+
+
+def _add_directory_to_tarfile(
+    tar_file: tarfile.TarFile,
+    origin_path: Path,
+    arcname: str = ".",
+) -> None:
+    """Add a directory to a tarfile."""
+    for directory_item in origin_path.iterdir():
+        arcpath = Path(arcname, directory_item.name).as_posix()
+
+        if directory_item.is_file():
+            if any(directory_item.match(exclude) for exclude in EXCLUDE_FROM_BACKUP):
+                continue
+            tar_file.add(directory_item.as_posix(), arcname=arcpath, recursive=False)
+            continue
+
+        if directory_item.is_dir():
+            _add_directory_to_tarfile(tar_file, directory_item, arcpath)
+            continue
+
+        tar_file.add(directory_item.as_posix(), arcname=arcpath, recursive=False)
```

Comment:
This is for symlinks right? Can we add a comment
```suggestion
        # Not a dir or file, so a symlink.
        tar_file.add(directory_item.as_posix(), arcname=arcpath, recursive=False)
```